### PR TITLE
Update Registry documentation according to SE-0292 amendment

### PR DESCRIPTION
**Pull request by mattt**
This PR makes the changes to the registry service specification described in https://github.com/apple/swift-evolution/pull/1410

### CODE REVIEW

**Review by neonichu — APPROVED** [review]

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: Documentation/Registry.md [hunk 1]
```diff
@@ -256,19 +253,15 @@ Package scopes are case-insensitive
 #### 3.6.2 Package name
 
 A package's *name* uniquely identifies a package in a scope.
-The maximum length of a package name is 128 characters.
+The maximum length of a package name is 100 characters.
 A valid package name matches the following regular expression pattern:
 
 ```regexp
-\A\p{XID_Start}\p{XID_Continue}{0,127}\z
+\A[a-zA-Z\d_.-]{1,100}\z
```
Comment (yim-lee) [inline Documentation/Registry.md:260]: Please see https://github.com/apple/swift-evolution/pull/1410#discussion_r673370637. 

I can see names beginning with `_` and `.` but not so much with `-`. Want to see what others think.

**Review by yim-lee — COMMENTED** [review]
https://github.com/apple/swift-package-manager/pull/3624/files#diff-1dfed3389cc482cc6446784dd9e800e5dbab6d62356bfac3010cc1382151c4bdR602-R603

> A client MAY send a `GET` request for a URI matching the expression `/{scope}/{name}/{version}`

I suggest changing the URI to `/{scope}/{name}/{version}.zip` since `.zip` is required.

File: Documentation/Registry.md (hunk 1 above)
Comment (mattt) [reply Documentation/Registry.md:260]: `.` is used as a separator between scope and name, so we probably shouldn't allow it as the first character (to avoid, for example, `mona.........LinkedList`.

**Comment by tomerd** [comment]
@mattt @yim-lee should we wait before merging these PRs to see if there are further amendments we need to make and then do all in one go?

File: Documentation/Registry.md [hunk 2]
```diff
@@ -661,23 +661,37 @@ as described by [RFC 6249].
...
+
+```http
+HTTP/1.1 200 OK
+Accept-Ranges: bytes
+Cache-Control: public, immutable
+Content-Type: application/zip
+Content-Disposition: attachment; filename="LinkedList-1.1.1.zip"
 Content-Length: 2048
+Content-Version: 1
 Digest: sha-256=a2ac54cf25fbc1ad0028f03f0aa4b96833b83bb05a14e510892bb27dea4dc812
```
Comment (yim-lee) [inline Documentation/Registry.md:692]: @mattt Diffing the latest changes against https://github.com/apple/swift-package-manager/pull/3624/commits/5e2b7acac36b4ff859adae00f768c6e5042ac974, are we going to keep `Digest` but remove `Location`?

Comment (mattt) [reply Documentation/Registry.md:692]: This example shows a sequence of:

1. A redirect response from the registry
2. A new request from the client using the `Location` of the redirect
3. A response from the new, redirected host

The first response includes a `Location` because it's a redirect, but the second response doesn't because it's serving the asset.

Comment (yim-lee) [reply Documentation/Registry.md:692]: Thanks, so this is proposing to remove `Digest` from the registry's redirect response.

There was a discussion on this in the past, and having `Digest` in the redirect response addresses a concern that @robertlacroix had: 

Screenshot [https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/5](https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/5):
image



Screenshot [https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/7](https://forums.swift.org/t/se-0292-3rd-review-package-registry-service/49107/7):
image

I want to get @robertlacroix's feedback here to make sure he is ok with this change.

Comment (mattt) [reply Documentation/Registry.md:692]: Copying my rationale from https://github.com/apple/swift-evolution/pull/1410 , for convenience: 

> **Redirection**: The current guidance to treat forwarding headers as authoritative is (as far as I can tell) unprecedented and possibly out-of-spec for HTTP. Our understanding of the feedback in https://github.com/apple/swift-evolution/pull/1319 is that the motivation for this change was to ensure that proxying services are supported. Conventional HTTP redirects support this use case without any special accommodations.

File: Documentation/Registry.md [hunk 3]
```diff
@@ -252,14 +252,17 @@ Package scopes are case-insensitive
 #### 3.6.2 Package name
 
 A package's *name* uniquely identifies a package in a scope.
+A package name consists of alphanumeric characters, underscores, and hyphens.
+Hyphens and underscores may not occur at the beginning or end,
+nor consecutively within a name.
 The maximum length of a package name is 100 characters.
-A valid package name matches the following regular expression pattern:
+A valid package scope matches the following regular expression pattern:
```
Comment (mattt) [inline Documentation/Registry.md:259]:
```suggestion
A valid package name matches the following regular expression pattern:
```

**Comment by mattt** [comment]
@swift-ci please smoke test

File: Documentation/Registry.md (hunk 2 above)
Comment (robertlacroix) [reply Documentation/Registry.md:695]: We added the digest to the package metadata returned by the registry, right? Therefore I'd prefer to remove examples of `Digest` headers from the responses serving the archive for two reasons:

1) There's a single source of truth (package metadata) and to be clear that clients must always use that as opposed to any `Digest` header
2) The archive might be served from a CDN that doesn't support that header

Comment (mattt) [reply Documentation/Registry.md:695]: > We added the digest to the package metadata returned by the registry, right? 

That was an idea that was suggested during 3rd review, but didn't make it into the final proposal.

> Therefore I'd prefer to remove examples of `Digest` headers from the responses serving the archive for two reasons:
> 
> 1. There's a single source of truth (package metadata) and to be clear that clients must always use that as opposed to any `Digest` header

I agree that a single source of truth is preferable for client validation. Our design is for the `Digest` header to serve that role. An advantage to this approach is that not additional requests are needed to get the digest when downloading a source archive for a resolved package version.

> 2. The archive might be served from a CDN that doesn't support that header

I'm unaware of any CDN providers that don't support origin header. I've personally worked with AWS CloudFront, Azure, Cloudflare, and Fastly, and know for a fact that they all support sending custom response headers. I'm not aware of any that do this automatically based on content, but you can at least set a precomputed header on the resource. Skimming the docs of other providers, it looks like Alibaba, Akamai, and Google Cloud also support custom headers.

File: Documentation/Registry.md [hunk 4]
```diff
@@ -661,31 +673,45 @@ as described by [RFC 6249].
...
 ### 4.5 Lookup package identifiers registered for a URL
 
 A client MAY send a `GET` request
 for a URI matching the expression `/identifiers{?url}`
 to retrieve package identifiers associated with a particular URL.
-A client SHOULD set the `Accept` header to
+A client SHOULD set the `Accept` header with the value
 `application/vnd.swift.registry.v1+json`.
 
 ```http
```
Comment (yim-lee) [inline Documentation/Registry.md:717]: The response for this API is an array of identifier strings:

```
{
    "identifiers": [
      "mona.LinkedList"
    ]
}
```

Would it better if it were an array of objects?

```
{
  "identifiers": [
    { 
      "scope": "mona",
      "name": "LinkedList"
    }
  ]
}
```

Comment (mattt) [reply Documentation/Registry.md:717]: The purpose of this endpoint is to give the registry a way to map existing URL-based dependencies with identifiers so that SwiftPM can resolve a mixed dependency graph.

In the case of registry configuration (`.swiftpm/config/registries.json`), it made sense to have an object (`"[default]": {"url": "https://..."}`) instead of a single value (`"[default]": "https://"`) because that would provide an extension point if we needed additional information later, like for authentication.

Here, I can't think of anything else we'd need to convey beyond the identifier. Decomposing the identifier into fields doesn't strike me as particularly necessary. Can you think of anything else we'd want to put here?

Comment (yim-lee) [reply Documentation/Registry.md:717]: I am trying to consider how the received data will be used. It's fine to keep identifiers as strings if we are just going to use them as-is, but if we ever need to extract scope and/or name then it might be more convenient to represent them as objects instead.

File: Documentation/Registry.md (hunk 2 above)
Comment (robertlacroix) [reply Documentation/Registry.md:704]: I would really like the registry to be in control of the authenticity of the package. As it's designed right now, a redirect on the `/{scope}/{name}/{version}.zip` endpoint yields control of the checksum to another server, which adds opportunity to maliciously alter the package.

Therefore I propose adding the digest to `/{scope}/{name}` and be that the source of truth. I agree it requires another request in some cases.

Comment (tomerd) [reply Documentation/Registry.md:704]: seems like a reasonable request to add this to the versions list API which SwiftPM hits anyways as part of the dependency resolution / version determination flow.

Comment (mattt) [reply Documentation/Registry.md:753]: @robertlacroix Thanks for weighing in on this. I just pushed 362099ae, which updates the response for `GET /{scope}/{name}/{version}` to include a `checksum` field that serves as the authoritative integrity check for package resources. This change also provides a more predictable, extensible structure to `GET /{scope}/{name}/{version}`. It also fixes a mistake that I didn't realize until now, which is that the HTTP `Digest` field defines `sha-256` to use a Base64 encoding rather than the hexadecimal representation used by `swift package generate-checksum`.

Comment (yim-lee) [reply Documentation/Registry.md:753]: Should `checksum` be added to both `GET /{scope}/{name}` and `GET /{scope}/{name}/{version}`? SwiftPM will hit the list package releases API more frequently I would think and having `checksum` in that response should serve as an optimization.

Comment (mattt) [reply Documentation/Registry.md:753]: @yim-lee I think the benefit of having a single source of truth outweighs the potential drawback of an extra HTTP request. And if we ever expand what resources are available (for binary distribution(s), software bill of materials, documentation archive, etc.), it'll be better to have those all together under `GET /{scope}/{name}/{version}`, rather than congesting the list of versions in `GET /{scope}/{name}`.

File: Documentation/Registry.md [hunk 5]
```diff
@@ -429,11 +433,32 @@ Otherwise, a server SHOULD respond with a status code of `404` (Not Found).
...
 Link: <https://packages.example.com/mona/LinkedList/1.1.1>; rel="latest-version",
       <https://packages.example.com/mona/LinkedList/1.0.0>; rel="predecessor-version"
+{
+  "id": "mona.LinkedList",
+  "version": "1.1.1",
+  "resources": [
+    {
+      "name": "source-archive",
+      "type": "application/zip",
+      "checksum": "a2ac54cf25fbc1ad0028f03f0aa4b96833b83bb05a14e510892bb27dea4dc812"
```
Comment (yim-lee) [inline Documentation/Registry.md:446]: Should this be `sha-256=a2ac54cf25fbc1ad0028f03f0aa4b96833b83bb05a14e510892bb27dea4dc812`?

Comment (mattt) [reply Documentation/Registry.md:446]: Now that we aren't using the standard `Digest` header, we don't need to specify an algorithm here. The `checksum` for a package is tied to the output of `swift package generate-checksum`, which is fixed to SHA256. I don't personally expect that we'll need to change this, but if we wanted to future-proof this, we could change this to:

```suggestion
      "checksum": {"sha-256": "a2ac54cf25fbc1ad0028f03f0aa4b96833b83bb05a14e510892bb27dea4dc812"}
```

File: Documentation/Registry.md [hunk 6]
```diff
@@ -647,10 +706,15 @@ and caching as described by [RFC 7234].
 
 #### 4.4.1. Integrity verification
 
-A client MUST verify the integrity of a downloaded source archive
-using the checksum provided in the `Digest` header of a response
-(for example, using the command
-`echo "$CHECKSUM LinkedList-1.1.1.zip" | shasum -a 256 -c`).
+A client MUST verify the integrity of a downloaded source archive using
+the `checksum` value for the associated `source-archive` resource
+in the response to `GET /{scope}/{name}/{version}`,
```
Comment (mattt) [inline Documentation/Registry.md:711]:
```suggestion
in the response to `GET /{scope}/{name}/{version}.zip`,
```

File: Documentation/Registry.md [hunk 7]
```diff
@@ -596,14 +655,14 @@ Location: https://packages.example.com/mona/LinkedList/1.1.1/Package.swift
 
 <a name="endpoint-4"></a>
 
-### 4.4. Fetch source archive
+### 4.4. Download source archive
 
 A client MAY send a `GET` request
 for a URI matching the expression `/{scope}/{name}/{version}`
```
Comment (mattt) [inline Documentation/Registry.md:661]:
```suggestion
for a URI matching the expression `/{scope}/{name}/{version}.zip`
```

**Comment by mattt** [comment]
@swift-ci Please smoke test

**Comment by tomerd** [comment]
@yim-lee @mattt do you feel this is ready to push forward with? We would probably need a short amendment review with the community

**Comment by tomerd** [comment]
@mattt @yim-lee amendment review: https://forums.swift.org/t/amendment-se-0292-package-registry-service/51663

File: Sources/SourceControl/GitRepository.swift [hunk 8]
```diff
@@ -482,6 +482,7 @@ public final class GitRepository: Repository, WorkingCheckout {
         try self.lock.withLock {
             try callGit("archive",
                         "--format", "zip",
+                        "--prefix", path.basenameWithoutExt,
```
Comment (yim-lee) [inline Sources/SourceControl/GitRepository.swift:485]: Should this change be in a separate PR?

File: Documentation/Registry.md [hunk 9]
```diff
@@ -536,14 +586,23 @@ the name of the manifest file
...
+with a value for each version-specific package manifest file
+in the release's source archive,
 whose filename matches the following regular expression pattern:
 
 ```regexp
-\APackage(?:@swift-\d+(?:\.\d+){0,2})?.swift\z
+\APackage@swift-(\d+)(?:\.(\d+))?(?:\.(\d+))?.swift\z
 ```
 
+Each link value SHOULD have the `alternative` relation type,
```
Comment (yim-lee) [inline Documentation/Registry.md:598]:
```suggestion
Each link value SHOULD have the `alternate` relation type,
```

**Comment by tomerd** [comment]
hi @mattt, with the amendment accepted I think we can now merge this. would you like to resolve the remaining conflicts, so I can take care of that?

**Comment by mattt** [comment]
@swift-ci Please smoke test

**Comment by mattt** [comment]
@tomerd Sure thing. That should be all of the pending feedback. Let me know if there's anything else you'd like me to do with this PR.
